README
Gregorian Calendar Functions
$Id: README,v 1.1 2005/04/20 18:47:35 gsotirov Exp $

---

This is one of my works when I was in Technical University of Sofia. Here I
implement some useful functions for work with the Gregorian calendar. I code
some algorithms for astronomical calculations and also wrote leap year checking
function.



Changelog
CHANGELOG

0.3.1 (2003-05-25)
-------------------------------------------------------------------------------
 - The name of CatholicEasterDay is wrong in the header file (fixed).

0.3.0 (2003-05-04)
-------------------------------------------------------------------------------
 - Some dummy functions removed
 - There are now two separate routines for calculation of the Catholic Easter day
   and Orthodox Easter day

0.1.0 (2001-12-14)
-------------------------------------------------------------------------------
 - Initial release



easter.c
/* File       : easter.c
 * Description: A test program for the Gregorian Calendar Functions.
 * Written by : Georgi D. Sotirov <gdsotirov@dir.bg>
 * CVS Id     : $Id: easter.c,v 1.3 2017/10/16 12:40:13 gsotirov Exp $
 */

#include <stdio.h>
#include <stdlib.h>
#include <unistd.h>

#include "gregor.h"

char *months[] = {
  "",
  "January",
  "February",
  "March",
  "April",
  "May",
  "June",
  "July",
  "August",
  "September",
  "October",
  "November",
  "December"};

int handleRequest(int year, int opt) {
  unsigned short month, day;

  if ( !isGregorian(year) ) {
    printf("Error: '%d' is not a valid Gregorian calendar year (between %d and %d)!\n", year, GREGORIAN_MIN, GREGORIAN_MAX);
    return 1;
  }

  switch (opt) {
    case 'c':
      CatholicEasterDay(&month, &day, year);
      printf("Catholic Easter day is on %d %s %d.\n", day, months[month], year);
      break;
    case 'l':
      if ( isLeapYear(year) ) {
        printf("Year %d is a leap year.\n", year);
      }
      else {
        printf("Year %d is not a leap year.\n", year);
      }
      break;
    case 'o':
      EasterDay(&month, &day, year);
      printf("Orthodox Easter day is on %d %s %d.\n", day, months[month], year);
      break;
  }

  return 0;
}

int main(int argc, char *argv[]) {
  int opt = 0;
  int year = 0;

  while ( (opt = getopt(argc, argv, "c:l:o:")) != -1 ) {
    switch (opt) {
      case 'c':
      case 'l':
      case 'o':
        year = atoi(optarg);
        handleRequest(year, opt);
        break;
      default:
        printf("Usage: %s -c <year> -l <year> -o <year>\n", argv[0]);
        printf("  -c <year> Calculate and print Catholic Easter date for given year.\n");
        printf("  -l <year> Check and print whether the given year is leap or not.\n");
        printf("  -o <year> Calculate and print Orthodox Easter date for given year.\n");
        printf("       %s <year>\n", argv[0]);
        printf("  Print everything for given year.\n");
    }
  }

  if ( argc > 1 && (year = atoi(argv[1])) != 0) {
    year = atoi(argv[1]);
  }
  else {
    printf("Please, enter year: ");
    scanf("%d", &year);
  }

  if ( !isGregorian(year) ) {
    printf("Error: %d is not a valid Gregorian calendar year (between %d and %d)!\n", year, GREGORIAN_MIN, GREGORIAN_MAX);
  }

  handleRequest(year, 'c');
  handleRequest(year, 'l');
  handleRequest(year, 'o');

  return 0;
}



gregor.h
/******************************************************************************
 *                        Gregorian Calendar Functions                        *
 ******************************************************************************
 * File         : gregor.h                                                    *
 * Version      : 0.3.1                                                       *
 * Date Started : 2001-12-14 12:00 GMT                                        *
 * Date Modified: 2003-05-04 14:00 GMT                                        *
 * Author       : Georgi D. Sotirov <gdsotirov@dir.bg>                        *
 * CVS Id       : $Id: gregor.h,v 1.1 2005/04/20 18:47:35 gsotirov Exp $
 ******************************************************************************/

#define GREGORIAN_MIN                   1582
#define GREGORIAN_MAX                   4092

#define EASTER_START_DATE_DAY           3
#define EASTER_START_DATE_MONTH         4 /* April */

int isGregorian(unsigned short year);
int isLeapYear(unsigned short year);
int CatholicEasterDay(unsigned short *month, unsigned short *day, unsigned short year);
int EasterDay(unsigned short *month, unsigned short *day, unsigned short year);



gregor.c
/******************************************************************************
 *                        Gregorian Calendar Functions                        *
 ******************************************************************************
 * File         : gregor.c                                                    *
 * Version      : 0.3.1                                                       *
 * Date started : 2001-12-14 12:00 GMT                                        *
 * Date modified: 2003-05-04 14:00 GMT                                        *
 * Author       : Georgi D. Sotirov <gdsotirov@dir.bg>                        *
 * CVS Id       : $Id: gregor.c,v 1.1 2005/04/20 18:47:35 gsotirov Exp $
 ******************************************************************************/

#include "gregor.h"

/******************************************************************************
 * Function     : isGregorian                                                 *
 * Purpose      : To find if a specific year is a gregorian calendar year.    *
 * Returns      : 1 if year is gregorian, and 0 if not                        *
 ******************************************************************************/

int isGregorian(year)
    unsigned short year; /* specific year */
{
    if ( year >= GREGORIAN_MIN && year <= GREGORIAN_MAX )
        return 1;
    return 0;
}

/******************************************************************************
 * Function     : isLeapYear                                                  *
 * Purpose      : Finds whether specific year is leap or not.                 *
 * Returns      : 1 if the year is leap or 0 otherwise.                       *
 ******************************************************************************/

int isLeapYear(year)
    unsigned short year;    /* specific year */
{
    if ( isGregorian(year) )
        return ( (year % 4 == 0) && ((year / 100 != 0) || (year % 400 == 0)) );
    return 0;
}

/******************************************************************************
 * Function     : CatholicEasterDay                                           *
 * Purpose      : Finds the date of catholic easter day with the J. M. Oudin  *
 *                algorithm (1940).                                           *
 * Returns      : One on successful operation otherwise zero.                 *
 ******************************************************************************

 Notes:
       See Meeus, "Astronomical Algorithms", p 67.  He  in  turn  states  that
    the  following  method  has  been  given  by  Spencer  Jones  in  his book
    "General   Astronomy"   (p 73-74  of  the  1922  edition).  It   has  been
    published again in  the "Journal of the British Astronomical Association",
    vol 88, page 91 (December 1977),  where it is said that  it was devised in
    1876 and appeared in Butcher's "Astronomical Calendar".
       Unlike  the  formula  given  by  Gauss,  this method has  no  exception
    and is valid for all years in the Gregorian calendar,  hence from 1583 on.
*/

int CatholicEasterDay(month, day, year)
    unsigned short *month;  /* month of easter */
    unsigned short *day;    /* day of easter */
    unsigned short year;    /* specific year */
{
    if ( isGregorian(year) ) {
        int c; /* algorithm variables */
        int n;
        int k;
        int i;
        int j;
        int l;

        c = year / 100;
        n = year - 19 * (year / 19);
        k = (c - 17) / 25;

        i = c - c / 4 - (c - k) / 3 + 19 * n + 15;
        i = i - 30 * (i / 30);
        i = i - (i / 28) * (1 - (i / 28) * (29 / (i + 1)) * ((21 - n) / 11));

        j = year + year / 4 + i + 2 - c + c / 4;
        j = j - 7 * (j / 7);

        l = i - j;

        *month = 3 + (l + 40) / 44;
        *day   = l + 28 - 31 * (*month / 4);
    }
    else {
        *month = *day = 0;
        return 0;
    }

    return 1;
}

/******************************************************************************
 * Function     : EasterDay                                                   *
 * Purpose      : Finds the date of easter day with the modified Gauss        *
 *                algorithm.                                                  *
 * Returns      : One on successful operation otherwise zero.                 *
 ******************************************************************************

 Notes:
    See "Universe", 2000, book 2, p 30 and the correction in "Universe", 2000,
    book 3, p 37.
*/

int EasterDay(month, day, year)
    unsigned short *month;
    unsigned short *day;
    unsigned short year;
{
    if ( isGregorian(year) ) {
        int n1; /* algorithm variables */
        int n2;
        int n3;
        int n4;
        int n5;
        int na;
        int nb;
        int nc;

        n1 = year % 19;
        n2 = year % 4;
        n3 = year % 7;

        na = 19 * n1 + 16;
        n4 = na % 30;
        nb = 2 * n2 + 4 * n3 + 6 * n4;
        n5 = nb % 7;
        nc = n4 + n5;

        *month = EASTER_START_DATE_MONTH;

        if ( nc > 30 - EASTER_START_DATE_DAY ) {
            *day = nc - (30 - EASTER_START_DATE_DAY);
            *month = *month + 1;
        }
        else *day = EASTER_START_DATE_DAY + nc;
    }
    else {
        *month = *day = 0;
        return 0;
    }

    return 1;
}

